{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from clusterlogs import pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/error0101.txt')\n",
    "#df = pd.read_csv('data/error0102.txt')\n",
    "#df = pd.read_csv('data/error0115.txt')\n",
    "#df = pd.read_csv('data/error0201.txt')\n",
    "#df = pd.read_csv('data/error0225.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>error_message</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [3010] Write permission denied</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>globus_ftp_client: the server responded with an error 500 Command failed : stat64 error: No such file or directory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                      error_message\n",
       "0  globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.   \n",
       "1  Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [3010] Write permission denied                                                                                      \n",
       "2  globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.   \n",
       "3  globus_ftp_client: the server responded with an error 500 Command failed : stat64 error: No such file or directory                                                                              \n",
       "4  globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.   "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(703546, 1)"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster = pipeline.Chain(df, target='error_message', mode='create')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tokenization finished\n",
      "Found 21755 equal groups\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ntuckus/anaconda3/envs/test/lib/python3.6/site-packages/smart_open/smart_open_lib.py:402: UserWarning: This function is deprecated, use smart_open.open instead. See the migration notes for details: https://github.com/RaRe-Technologies/smart_open/blob/master/README.rst#migrating-to-the-new-open-function\n",
      "  'See the migration notes for details: %s' % _MIGRATION_NOTES_URL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vectorization of tokens finished\n",
      "Vectorization of sentences is finished\n",
      "Vocabulary size = 18847\n",
      "Number of dimensions is 137\n",
      "DBSCAN finished with 391 clusters\n",
      "Matching Clusterization!\n",
      "Postprocessed with 281 clusters\n"
     ]
    }
   ],
   "source": [
    "cluster.process()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pattern</th>\n",
       "      <th>indices</th>\n",
       "      <th>cluster_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>err: Communication error on send, err: [SE][][] httpg:/:/srm/: CGSI-gSOAP running on reports:</td>\n",
       "      <td>[662665, 561309, 580151, 524045, 671485, 355154, 227099, 340623, 403862, 541966, 13530, 13650, 13721, 28837, 82011, 116632, 117707, 117710, 117957, 131531, 131532, 131548, 131570, 131585, 131601, 142510, 142543, 150508, 152794, 152799, 153010, 164599, 164601, 164653, 164665, 164672, 170366, 174650, 174670, 176826, 176957, 201634, 201650, 201652, 201688, 201689, 201691, 201693, 202223, 202226, 202279, 202286, 202290, 202302, 202843, 202885, 294269, 294295, 296543, 296551, 296592, 296615, 296662, 297496, 298614, 298617, 298618, 298937, 299448, 299586, 378552, 403823, 404018, 404020, 406347, 406379, 422191, 422197, 422225, 422323, 422350, 431666, 431899, 444109, 447461, 447742, 447744, 517957, 517974, 517978, 518016, 518027, 518528, 518529, 518548, 518587, 519127, 519169, 541421, 541434, ...]</td>\n",
       "      <td>100544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []</td>\n",
       "      <td>[1, 18, 83, 93, 110, 133, 184, 274, 287, 289, 297, 304, 359, 360, 394, 397, 400, 406, 429, 434, 439, 442, 450, 457, 461, 463, 465, 466, 481, 490, 495, 511, 525, 527, 533, 535, 549, 573, 577, 579, 591, 592, 622, 641, 670, 674, 694, 715, 722, 726, 763, 770, 809, 841, 844, 875, 879, 887, 888, 901, 909, 911, 937, 939, 960, 965, 970, 980, 1004, 1010, 1020, 1048, 1054, 1095, 1098, 1122, 1138, 1144, 1145, 1167, 1170, 1175, 1185, 1217, 1240, 1257, 1272, 1273, 1297, 1300, 1316, 1331, 1337, 1346, 1351, 1355, 1361, 1373, 1380, 1411, ...]</td>\n",
       "      <td>60088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094//atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>[41310, 41369, 41437, 41522, 101955, 103616, 104196, 104261, 104503, 186502, 197847, 266042, 331741, 331947, 332055, 344237, 357650, 357691, 361859, 366514, 366630, 471010, 471216, 471482, 473303, 473432, 473909, 485400, 517221, 571724, 571725, 592146, 615007, 615395, 615520, 615734, 643412, 643479, 669270, 31410, 40735, 40980, 41311, 97108, 101752, 103097, 103145, 128048, 162314, 167669, 168062, 181775, 184079, 185623, 268117, 286862, 291776, 322521, 322702, 357169, 357326, 389365, 428490, 428821, 470485, 484978, 547813, 548070, 571691, 573767, 589364, 589435, 667350, 667375, 668771, 56837, 65602, 91981, 92138, 92278, 92531, 211305, 211397, 269804, 270706, 270801, 270848, 370773, 418732, 451406, 551348, 551587, 576451, 674997, 675240, 679371, 679468, 681219, 684654, 684658, ...]</td>\n",
       "      <td>45044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [4003] Authentication to:1094 failed; all protocols have been tried.</td>\n",
       "      <td>[277, 769, 969, 1947, 2287, 2330, 2352, 4173, 4619, 4847, 5950, 6320, 7429, 7795, 7887, 8942, 9047, 9722, 9723, 9853, 10126, 10241, 11778, 11877, 12379, 13349, 13819, 13979, 14089, 14713, 14924, 14980, 15698, 15970, 17107, 17843, 19349, 20843, 21317, 21675, 22112, 23194, 23807, 25268, 25312, 26215, 26399, 26446, 28278, 29348, 29682, 29725, 29785, 29818, 30003, 31567, 31775, 32991, 34953, 35372, 35490, 37490, 39886, 40679, 40781, 41326, 41475, 41619, 41988, 42412, 42530, 42544, 43188, 45374, 45457, 45565, 45911, 46504, 47869, 48304, 48445, 48451, 48584, 48896, 49143, 49539, 50290, 50810, 52137, 52497, 52800, 53657, 56561, 56587, 56685, 57398, 57536, 59408, 59450, 59946, ...]</td>\n",
       "      <td>29732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>: the server responded with an error 500 500-Command failed : System error in: No such file or directory 500-A system call failed: No such file or directory 500 End.</td>\n",
       "      <td>[31917, 31921, 31979, 32562, 41353, 166551, 168232, 170116, 170201, 186321, 471127, 471227, 471771, 473453, 476462, 486004, 557910, 609548, 634215, 659854, 560, 629, 871, 872, 1651, 1901, 2142, 2293, 2298, 2458, 2549, 2799, 2974, 2980, 3079, 3082, 3167, 3512, 3513, 3712, 3747, 3748, 3790, 3793, 3925, 4012, 4136, 4188, 5056, 5061, 5131, 5245, 5252, 5511, 5683, 5684, 5851, 5921, 6027, 6098, 6349, 6541, 6616, 6617, 6719, 6798, 6971, 6995, 7109, 7232, 7245, 7307, 7576, 7664, 7668, 8073, 8169, 8225, 8226, 8356, 8444, 8458, 8609, 8650, 8700, 8763, 8846, 9020, 9022, 9072, 9139, 9145, 9202, 9335, 9339, 9358, 9414, 9710, 9776, 9865, ...]</td>\n",
       "      <td>29278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                 pattern  \\\n",
       "0  err: Communication error on send, err: [SE][][] httpg:/:/srm/: CGSI-gSOAP running on reports:                                                                           \n",
       "1  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []                                                                                            \n",
       "7  Protocol not supported or path/url invalid: root:/:1094//atlasscratchdisk/rucio/tests///                                                                                \n",
       "2  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [4003] Authentication to:1094 failed; all protocols have been tried.                          \n",
       "3  : the server responded with an error 500 500-Command failed : System error in: No such file or directory 500-A system call failed: No such file or directory 500 End.   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            indices  \\\n",
       "0  [662665, 561309, 580151, 524045, 671485, 355154, 227099, 340623, 403862, 541966, 13530, 13650, 13721, 28837, 82011, 116632, 117707, 117710, 117957, 131531, 131532, 131548, 131570, 131585, 131601, 142510, 142543, 150508, 152794, 152799, 153010, 164599, 164601, 164653, 164665, 164672, 170366, 174650, 174670, 176826, 176957, 201634, 201650, 201652, 201688, 201689, 201691, 201693, 202223, 202226, 202279, 202286, 202290, 202302, 202843, 202885, 294269, 294295, 296543, 296551, 296592, 296615, 296662, 297496, 298614, 298617, 298618, 298937, 299448, 299586, 378552, 403823, 404018, 404020, 406347, 406379, 422191, 422197, 422225, 422323, 422350, 431666, 431899, 444109, 447461, 447742, 447744, 517957, 517974, 517978, 518016, 518027, 518528, 518529, 518548, 518587, 519127, 519169, 541421, 541434, ...]   \n",
       "1  [1, 18, 83, 93, 110, 133, 184, 274, 287, 289, 297, 304, 359, 360, 394, 397, 400, 406, 429, 434, 439, 442, 450, 457, 461, 463, 465, 466, 481, 490, 495, 511, 525, 527, 533, 535, 549, 573, 577, 579, 591, 592, 622, 641, 670, 674, 694, 715, 722, 726, 763, 770, 809, 841, 844, 875, 879, 887, 888, 901, 909, 911, 937, 939, 960, 965, 970, 980, 1004, 1010, 1020, 1048, 1054, 1095, 1098, 1122, 1138, 1144, 1145, 1167, 1170, 1175, 1185, 1217, 1240, 1257, 1272, 1273, 1297, 1300, 1316, 1331, 1337, 1346, 1351, 1355, 1361, 1373, 1380, 1411, ...]                                                                                                                                                                                                                                                                               \n",
       "7  [41310, 41369, 41437, 41522, 101955, 103616, 104196, 104261, 104503, 186502, 197847, 266042, 331741, 331947, 332055, 344237, 357650, 357691, 361859, 366514, 366630, 471010, 471216, 471482, 473303, 473432, 473909, 485400, 517221, 571724, 571725, 592146, 615007, 615395, 615520, 615734, 643412, 643479, 669270, 31410, 40735, 40980, 41311, 97108, 101752, 103097, 103145, 128048, 162314, 167669, 168062, 181775, 184079, 185623, 268117, 286862, 291776, 322521, 322702, 357169, 357326, 389365, 428490, 428821, 470485, 484978, 547813, 548070, 571691, 573767, 589364, 589435, 667350, 667375, 668771, 56837, 65602, 91981, 92138, 92278, 92531, 211305, 211397, 269804, 270706, 270801, 270848, 370773, 418732, 451406, 551348, 551587, 576451, 674997, 675240, 679371, 679468, 681219, 684654, 684658, ...]             \n",
       "2  [277, 769, 969, 1947, 2287, 2330, 2352, 4173, 4619, 4847, 5950, 6320, 7429, 7795, 7887, 8942, 9047, 9722, 9723, 9853, 10126, 10241, 11778, 11877, 12379, 13349, 13819, 13979, 14089, 14713, 14924, 14980, 15698, 15970, 17107, 17843, 19349, 20843, 21317, 21675, 22112, 23194, 23807, 25268, 25312, 26215, 26399, 26446, 28278, 29348, 29682, 29725, 29785, 29818, 30003, 31567, 31775, 32991, 34953, 35372, 35490, 37490, 39886, 40679, 40781, 41326, 41475, 41619, 41988, 42412, 42530, 42544, 43188, 45374, 45457, 45565, 45911, 46504, 47869, 48304, 48445, 48451, 48584, 48896, 49143, 49539, 50290, 50810, 52137, 52497, 52800, 53657, 56561, 56587, 56685, 57398, 57536, 59408, 59450, 59946, ...]                                                                                                                         \n",
       "3  [31917, 31921, 31979, 32562, 41353, 166551, 168232, 170116, 170201, 186321, 471127, 471227, 471771, 473453, 476462, 486004, 557910, 609548, 634215, 659854, 560, 629, 871, 872, 1651, 1901, 2142, 2293, 2298, 2458, 2549, 2799, 2974, 2980, 3079, 3082, 3167, 3512, 3513, 3712, 3747, 3748, 3790, 3793, 3925, 4012, 4136, 4188, 5056, 5061, 5131, 5245, 5252, 5511, 5683, 5684, 5851, 5921, 6027, 6098, 6349, 6541, 6616, 6617, 6719, 6798, 6971, 6995, 7109, 7232, 7245, 7307, 7576, 7664, 7668, 8073, 8169, 8225, 8226, 8356, 8444, 8458, 8609, 8650, 8700, 8763, 8846, 9020, 9022, 9072, 9139, 9145, 9202, 9335, 9339, 9358, 9414, 9710, 9776, 9865, ...]                                                                                                                                                                       \n",
       "\n",
       "   cluster_size  \n",
       "0  100544        \n",
       "1  60088         \n",
       "7  45044         \n",
       "2  29732         \n",
       "3  29278         "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#some of the clusters\n",
    "cluster.result.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [4003] Authentication to agh3.atlas.unimelb.edu.au:1094 failed; all protocols have been tried. '\n",
      " 'Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [4003] Authentication to agh3.atlas.unimelb.edu.au:1094 failed; all protocols have been tried. '\n",
      " 'Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [4003] Authentication to agh3.atlas.unimelb.edu.au:1094 failed; all protocols have been tried. '\n",
      " ...\n",
      " 'Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [4003] Authentication to xrootd.aglt2.org:1094 failed; all protocols have been tried. '\n",
      " 'Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [4003] Authentication to xrootd.aglt2.org:1094 failed; all protocols have been tried. '\n",
      " 'Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [4003] Authentication to xrootd.aglt2.org:1094 failed; all protocols have been tried. '] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "#some messages from the cluster 2\n",
    "print(cluster.in_cluster(cluster.result, 2),\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cluster.result['pattern'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pattern</th>\n",
       "      <th>indices</th>\n",
       "      <th>cluster_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>226</td>\n",
       "      <td>TRANSFER｟*｠: Unable to connect to｟*｠:2811｟*｠: globus_xio_system_socket_register_connect failed.｟*｠: System error in connect: Network is unreachable｟*｠: A system call failed: Network is unreachable</td>\n",
       "      <td>[97643, 547992]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>234</td>\n",
       "      <td>TRANSFER globus_ftp_client: the server responded with an error 500 Cannot enter passive mode: java.net.SocketException: Invalid argument</td>\n",
       "      <td>[417216, 517896]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>227</td>\n",
       "      <td>[][][][｟*｠][｟*｠] Error reported from srm_ifce : 13 [SE][Ls][SRM_AUTHORIZATION_FAILURE] User does not have valid permissions</td>\n",
       "      <td>[376053, 36658]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>256</td>\n",
       "      <td>TRANSFER ERROR: Copy failed with mode 3rd push, with error: Transfer failed: failure: Remote copy failed with status code 0: Failed to connect to｟*｠ 443:</td>\n",
       "      <td>[539911, 440066]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>235</td>\n",
       "      <td>DESTINATION EXISTS Destination already｟*｠ gsiftp:/eoscmsftp.cern.ch/eos/cms/store/mc///GEN/106X_upgrade2018_realistic_v4-v1//, Cancel</td>\n",
       "      <td>[424553, 316501]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>237</td>\n",
       "      <td>DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][SRM_INTERNAL_ERROR] Request to [&gt;PnfsManager@local] timed out.</td>\n",
       "      <td>[621859, 691305]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>233</td>\n",
       "      <td>DESTINATION CHECKSUM MISMATCH Source and destination ADLER32 do not match (｟*｠ !=｟*｠)</td>\n",
       "      <td>[101114, 234064]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>232</td>\n",
       "      <td>TRANSFER ERROR: Copy failed with mode 3rd push, with error: Transfer failed: failure: failed to connect to server: javax.net.ssl.SSLPeerUnverifiedException: Certificate for &lt;dynafed02.heprc.uvic.ca&gt; doesnt match any of the subject alternative names: [mebert@uvic.ca]</td>\n",
       "      <td>[448660, 465342]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>231</td>\n",
       "      <td>DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][SRM_INTERNAL_ERROR] error accessing stor110.cis.gov.pl:/mnt/dpm/cms/cms/2020-01-01:｟*｠ out</td>\n",
       "      <td>[397923, 441065]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>230</td>\n",
       "      <td>SOURCE SRM_GET_TURL srm-ifce err: Connection reset by peer, err: [SE][｟*｠][] httpg:/srm-eoslhcb.cern.ch:8443/srm/v2/server: CGSI-gSOAP running on reports Error reading token data header: Connection reset by peer</td>\n",
       "      <td>[527149, 271286]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                        pattern  \\\n",
       "226  TRANSFER｟*｠: Unable to connect to｟*｠:2811｟*｠: globus_xio_system_socket_register_connect failed.｟*｠: System error in connect: Network is unreachable｟*｠: A system call failed: Network is unreachable                                                                         \n",
       "234  TRANSFER globus_ftp_client: the server responded with an error 500 Cannot enter passive mode: java.net.SocketException: Invalid argument                                                                                                                                     \n",
       "227  [][][][｟*｠][｟*｠] Error reported from srm_ifce : 13 [SE][Ls][SRM_AUTHORIZATION_FAILURE] User does not have valid permissions                                                                                                                                                  \n",
       "256  TRANSFER ERROR: Copy failed with mode 3rd push, with error: Transfer failed: failure: Remote copy failed with status code 0: Failed to connect to｟*｠ 443:                                                                                                                    \n",
       "235  DESTINATION EXISTS Destination already｟*｠ gsiftp:/eoscmsftp.cern.ch/eos/cms/store/mc///GEN/106X_upgrade2018_realistic_v4-v1//, Cancel                                                                                                                                        \n",
       "237  DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][SRM_INTERNAL_ERROR] Request to [>PnfsManager@local] timed out.                                                                                                                                 \n",
       "233  DESTINATION CHECKSUM MISMATCH Source and destination ADLER32 do not match (｟*｠ !=｟*｠)                                                                                                                                                                                        \n",
       "232  TRANSFER ERROR: Copy failed with mode 3rd push, with error: Transfer failed: failure: failed to connect to server: javax.net.ssl.SSLPeerUnverifiedException: Certificate for <dynafed02.heprc.uvic.ca> doesnt match any of the subject alternative names: [mebert@uvic.ca]   \n",
       "231  DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][SRM_INTERNAL_ERROR] error accessing stor110.cis.gov.pl:/mnt/dpm/cms/cms/2020-01-01:｟*｠ out                                                                                                     \n",
       "230  SOURCE SRM_GET_TURL srm-ifce err: Connection reset by peer, err: [SE][｟*｠][] httpg:/srm-eoslhcb.cern.ch:8443/srm/v2/server: CGSI-gSOAP running on reports Error reading token data header: Connection reset by peer                                                          \n",
       "\n",
       "              indices  cluster_size  \n",
       "226  [97643, 547992]   2             \n",
       "234  [417216, 517896]  2             \n",
       "227  [376053, 36658]   2             \n",
       "256  [539911, 440066]  2             \n",
       "235  [424553, 316501]  2             \n",
       "237  [621859, 691305]  2             \n",
       "233  [101114, 234064]  2             \n",
       "232  [448660, 465342]  2             \n",
       "231  [397923, 441065]  2             \n",
       "230  [527149, 271286]  2             "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters, outliers = cluster.split_clusters(cluster.result, 'cluster_size', threshold=3)\n",
    "outliers.head(10) #clusters containing fewer than 3 messages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_size</th>\n",
       "      <th>pattern</th>\n",
       "      <th>mean_similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>100544</td>\n",
       "      <td>err: Communication error on send, err: [SE][][] httpg:/:/srm/: CGSI-gSOAP running on reports:</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>60088</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []</td>\n",
       "      <td>0.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>45044</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094//atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>29732</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [4003] Authentication to:1094 failed; all protocols have been tried.</td>\n",
       "      <td>0.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>29278</td>\n",
       "      <td>: the server responded with an error 500 500-Command failed : System error in: No such file or directory 500-A system call failed: No such file or directory 500 End.</td>\n",
       "      <td>0.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>27036</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [] Unable to ///////;</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>26537</td>\n",
       "      <td>Destination file exists and overwrite is not enabled</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>24267</td>\n",
       "      <td>｟*｠: Unable to connect to:2811｟*｠: System error in connect: Connection timed out｟*｠: A system call failed: Connection timed out</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>19802</td>\n",
       "      <td>Error reported from srm_ifce : 2 [SE][Ls][SRM_INVALID_PATH] No such file or directory</td>\n",
       "      <td>0.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>18333</td>\n",
       "      <td>end-of-file was reached｟*｠: An end of file occurred</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>18207</td>\n",
       "      <td>[gfalt_copy_file][perform_copy][gfal_xrootd_3rd_copy][gfal_xrootd_3rd_copy_bulk] Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []</td>\n",
       "      <td>0.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>17760</td>\n",
       "      <td>SOURCE SRM_GET_TURL error on the turl request : [SE][][SRM_FILE_UNAVAILABLE] File is.</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>16797</td>\n",
       "      <td>TRANSFER Transfer canceled because the gsiftp performance marker timeout of 360 seconds has been exceeded, or all performance markers during that period indicated zero bytes transferred</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>15173</td>\n",
       "      <td>on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Unable to open /dpm/nipne.ro/home/atlas/atlasscratchdisk/rucio/tests///; no delegated credentials for tpc</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>13091</td>\n",
       "      <td>: the server responded with an error 421 Service busy: Connection limit exceeded. Please try again later. Closing control connection.</td>\n",
       "      <td>0.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>12666</td>\n",
       "      <td>: the server responded with an error 530 530-globus_xio: Authentication Error 530-OpenSSL Error: s3_srvr.c:: in library: SSL routines, function: certificate 530-globus_gsi_callback_module: Could not verify credential 530-globus_gsi_callback_module: The certificate: with subject: /=/=/OU=/CN=/CN=/CN=/CN=/CN=</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>11878</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Access denied: /pnfs/triumf.ca/data/atlas/atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>11777</td>\n",
       "      <td>DESTINATION｟*｠ Connection timed out</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>11284</td>\n",
       "      <td>on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3012] sync - TPC open failed for src_url=root:/://atlasscratchdisk/rucio/tests///?fst.readahead=true</td>\n",
       "      <td>0.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>10277</td>\n",
       "      <td>TRANSFER globus_ftp_client: the server responded with an error 500 Command failed. : open/create : [ERROR] Server responded with an error: [3021] Unable to get quota space - quota not defined or exhausted /eos/cms/store/user/bmarzocc//94X_mc2017-RunIIFall18/////; Disk quota</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cluster_size  \\\n",
       "0   100544         \n",
       "1   60088          \n",
       "2   45044          \n",
       "3   29732          \n",
       "4   29278          \n",
       "5   27036          \n",
       "6   26537          \n",
       "7   24267          \n",
       "8   19802          \n",
       "9   18333          \n",
       "10  18207          \n",
       "11  17760          \n",
       "12  16797          \n",
       "13  15173          \n",
       "14  13091          \n",
       "15  12666          \n",
       "16  11878          \n",
       "17  11777          \n",
       "18  11284          \n",
       "19  10277          \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                 pattern  \\\n",
       "0   err: Communication error on send, err: [SE][][] httpg:/:/srm/: CGSI-gSOAP running on reports:                                                                                                                                                                                                                          \n",
       "1   Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []                                                                                                                                                                                                                                           \n",
       "2   Protocol not supported or path/url invalid: root:/:1094//atlasscratchdisk/rucio/tests///                                                                                                                                                                                                                               \n",
       "3   Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [4003] Authentication to:1094 failed; all protocols have been tried.                                                                                                                                                                         \n",
       "4   : the server responded with an error 500 500-Command failed : System error in: No such file or directory 500-A system call failed: No such file or directory 500 End.                                                                                                                                                  \n",
       "5   Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [] Unable to ///////;                                                                                                                                                                                                                        \n",
       "6   Destination file exists and overwrite is not enabled                                                                                                                                                                                                                                                                   \n",
       "7   ｟*｠: Unable to connect to:2811｟*｠: System error in connect: Connection timed out｟*｠: A system call failed: Connection timed out                                                                                                                                                                                        \n",
       "8   Error reported from srm_ifce : 2 [SE][Ls][SRM_INVALID_PATH] No such file or directory                                                                                                                                                                                                                                  \n",
       "9   end-of-file was reached｟*｠: An end of file occurred                                                                                                                                                                                                                                                                    \n",
       "10  [gfalt_copy_file][perform_copy][gfal_xrootd_3rd_copy][gfal_xrootd_3rd_copy_bulk] Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []                                                                                                                                                          \n",
       "11  SOURCE SRM_GET_TURL error on the turl request : [SE][][SRM_FILE_UNAVAILABLE] File is.                                                                                                                                                                                                                                  \n",
       "12  TRANSFER Transfer canceled because the gsiftp performance marker timeout of 360 seconds has been exceeded, or all performance markers during that period indicated zero bytes transferred                                                                                                                              \n",
       "13  on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Unable to open /dpm/nipne.ro/home/atlas/atlasscratchdisk/rucio/tests///; no delegated credentials for tpc                                                                                                                                   \n",
       "14  : the server responded with an error 421 Service busy: Connection limit exceeded. Please try again later. Closing control connection.                                                                                                                                                                                  \n",
       "15  : the server responded with an error 530 530-globus_xio: Authentication Error 530-OpenSSL Error: s3_srvr.c:: in library: SSL routines, function: certificate 530-globus_gsi_callback_module: Could not verify credential 530-globus_gsi_callback_module: The certificate: with subject: /=/=/OU=/CN=/CN=/CN=/CN=/CN=   \n",
       "16  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Access denied: /pnfs/triumf.ca/data/atlas/atlasscratchdisk/rucio/tests///                                                                                                                                                             \n",
       "17  DESTINATION｟*｠ Connection timed out                                                                                                                                                                                                                                                                                    \n",
       "18  on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3012] sync - TPC open failed for src_url=root:/://atlasscratchdisk/rucio/tests///?fst.readahead=true                                                                                                                                              \n",
       "19  TRANSFER globus_ftp_client: the server responded with an error 500 Command failed. : open/create : [ERROR] Server responded with an error: [3021] Unable to get quota space - quota not defined or exhausted /eos/cms/store/user/bmarzocc//94X_mc2017-RunIIFall18/////; Disk quota                                     \n",
       "\n",
       "    mean_similarity  \n",
       "0   0.72             \n",
       "1   0.99             \n",
       "2   0.62             \n",
       "3   0.88             \n",
       "4   0.94             \n",
       "5   0.81             \n",
       "6   1.00             \n",
       "7   0.84             \n",
       "8   0.99             \n",
       "9   1.00             \n",
       "10  0.99             \n",
       "11  1.00             \n",
       "12  1.00             \n",
       "13  0.85             \n",
       "14  0.98             \n",
       "15  0.86             \n",
       "16  0.83             \n",
       "17  1.00             \n",
       "18  0.75             \n",
       "19  0.93             "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#top 20 clusters by the size of cluster\n",
    "top_numb = 20\n",
    "top_clusters = cluster.validation(clusters)[['cluster_size','pattern','mean_similarity']].sort_values(by=[\"cluster_size\"], ascending=False) #clusters sorted by similarity inside the cluster\n",
    "top_clusters.head(top_numb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_size</th>\n",
       "      <th>pattern</th>\n",
       "      <th>mean_similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>110</td>\n",
       "      <td>155</td>\n",
       "      <td>｟*｠</td>\n",
       "      <td>0.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>38</td>\n",
       "      <td>3415</td>\n",
       "      <td>TRANSFER ERROR: Copy failed with mode 3rd push, with error:::::.</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>57</td>\n",
       "      <td>1259</td>\n",
       "      <td>TRANSFER CHECKSUM MISMATCH and｟*｠</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>118</td>\n",
       "      <td>112</td>\n",
       "      <td>: the server responded with an error 535 Authentication failed: The peers certificate with subjects DN CN====== was rejected. The peers certificate status is: FAILED The following validation errors were found: error at position in chain, problematic certificate subject: CN===</td>\n",
       "      <td>0.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>5400</td>\n",
       "      <td>TRANSFER ERROR: Copy failed with mode 3rd push, with error: copy Could not the proxy: fault: SOAP-ENV:Client [no subcode]: Detail:</td>\n",
       "      <td>0.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>151</td>\n",
       "      <td>28</td>\n",
       "      <td>DESTINATION SRM_PUT_TURL error on the turl request : [SE][][]</td>\n",
       "      <td>0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>153</td>\n",
       "      <td>24</td>\n",
       "      <td>TRANSFER ERROR: Copy failed with mode 3rd push, with error: Transfer failed: failure::; redirections [:/:///2020-01-01/?sfn=%2Fdpm%%2Fhome%%%%&amp;token=%</td>\n",
       "      <td>0.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35</td>\n",
       "      <td>3748</td>\n",
       "      <td>ERROR: Copy failed with mode 3rd push, with error: Transfer failed::</td>\n",
       "      <td>0.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>155</td>\n",
       "      <td>22</td>\n",
       "      <td>DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][SRM_INVALID_PATH] ///cms/store/////</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>45044</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094//atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster_size  \\\n",
       "110  155            \n",
       "38   3415           \n",
       "57   1259           \n",
       "118  112            \n",
       "30   5400           \n",
       "151  28             \n",
       "153  24             \n",
       "35   3748           \n",
       "155  22             \n",
       "2    45044          \n",
       "\n",
       "                                                                                                                                                                                                                                                                                  pattern  \\\n",
       "110  ｟*｠                                                                                                                                                                                                                                                                                    \n",
       "38   TRANSFER ERROR: Copy failed with mode 3rd push, with error:::::.                                                                                                                                                                                                                       \n",
       "57   TRANSFER CHECKSUM MISMATCH and｟*｠                                                                                                                                                                                                                                                      \n",
       "118  : the server responded with an error 535 Authentication failed: The peers certificate with subjects DN CN====== was rejected. The peers certificate status is: FAILED The following validation errors were found: error at position in chain, problematic certificate subject: CN===   \n",
       "30   TRANSFER ERROR: Copy failed with mode 3rd push, with error: copy Could not the proxy: fault: SOAP-ENV:Client [no subcode]: Detail:                                                                                                                                                     \n",
       "151  DESTINATION SRM_PUT_TURL error on the turl request : [SE][][]                                                                                                                                                                                                                          \n",
       "153  TRANSFER ERROR: Copy failed with mode 3rd push, with error: Transfer failed: failure::; redirections [:/:///2020-01-01/?sfn=%2Fdpm%%2Fhome%%%%&token=%                                                                                                                                 \n",
       "35   ERROR: Copy failed with mode 3rd push, with error: Transfer failed::                                                                                                                                                                                                                   \n",
       "155  DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][SRM_INVALID_PATH] ///cms/store/////                                                                                                                                                                      \n",
       "2    Protocol not supported or path/url invalid: root:/:1094//atlasscratchdisk/rucio/tests///                                                                                                                                                                                               \n",
       "\n",
       "     mean_similarity  \n",
       "110  0.24             \n",
       "38   0.26             \n",
       "57   0.50             \n",
       "118  0.52             \n",
       "30   0.52             \n",
       "151  0.54             \n",
       "153  0.55             \n",
       "35   0.59             \n",
       "155  0.61             \n",
       "2    0.62             "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#most inaccurate clusters\n",
    "top_clusters[['cluster_size','pattern','mean_similarity']].sort_values(by=[\"mean_similarity\"]).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'73.85% of all logs are in the top 20 of 281 clusters'"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "perc = round(top_clusters['cluster_size'].head(20).sum()/(df.shape[0]-1)*100,2)\n",
    "f\"{perc}% of all logs are in the top {top_numb} of {len(cluster.result)} clusters\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'tokenization': 305.0715,\n",
       " 'group_equals': 30.3863,\n",
       " 'tokens_vectorization': 2.7655,\n",
       " 'sentence_vectorization': 4.0631,\n",
       " 'ml_clusterization': 6.8031,\n",
       " 'matching_clusterization': 0.6947,\n",
       " 'process': 349.7874,\n",
       " 'validation': 8.4011}"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster.timings"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
